File: app/src/main/java/getyourcasts/jd/com/getyourcasts/repository/remote/RemoteDataRepository.java
package getyourcasts.jd.com.getyourcasts.repository.remote;

import android.content.ContentValues;
import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import getyourcasts.jd.com.getyourcasts.repository.remote.data.Channel;
import getyourcasts.jd.com.getyourcasts.repository.remote.data.Episode;
import getyourcasts.jd.com.getyourcasts.repository.remote.data.FeedResponse;
import getyourcasts.jd.com.getyourcasts.repository.remote.data.ItuneResponse;
import getyourcasts.jd.com.getyourcasts.repository.remote.data.Podcast;
import getyourcasts.jd.com.getyourcasts.repository.remote.network.NetworkHelper;
import okhttp3.Request;
import retrofit2.Call;

/**
 * Created by chuondao on 8/9/17.
 */

public class RemoteDataRepository implements DataRepository {


    private static final String TAG = "REMOTE_REPO";
    private static RemoteDataRepository instance = null;

    /**
     * getter for data instance
     */
    public static RemoteDataRepository getDataInstance() {
        if (instance == null) {
            instance =  new RemoteDataRepository();
        }
        return instance;
    }

    private RemoteDataRepository() {
    }

    @NotNull
    @Override
    public List<Podcast> searchPodcast(@NotNull String title) {
        List<Podcast> results = new ArrayList<>();
        Call<ItuneResponse> searchUrl = NetworkHelper.getHelperInstance().searchPodcast(title);
        Request http_req = searchUrl.request();
        Log.d(TAG, "Request "+http_req.toString());
        try {
            retrofit2.Response<ItuneResponse> response = searchUrl.execute();
        /*verify if response is good*/
            if (response.isSuccessful()) {
                // parse data here
                if (response.body() != null) return response.body().getResults();
                else {
                    Log.e(TAG, "Failed tor retreive message body");
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return results;
    }

    @Override
    public Channel fetchEpisodesFromFeedUrl(String feedUrl) {
        // get fetch RSS opt
        Call<FeedResponse> fetchRssOpt = NetworkHelper.getHelperInstance().fetchRss(feedUrl);

        if (fetchRssOpt != null) {
            Request req = fetchRssOpt.request();
            Log.d(TAG, "SEND REQUEST "+req.toString());
            try {
                retrofit2.Response<FeedResponse> feedResponse = fetchRssOpt.execute();
                if (feedResponse.isSuccessful()) {
                    if (feedResponse.body() != null) {
                        return feedResponse.body().getChannel();
                    } else {
                        Log.e(TAG, "Failed to fetch Media Item");
                    }
                }
            } catch(Exception e) {
                Log.e(TAG, "FAILED REQUEST $req. "+e.getMessage());
                e.printStackTrace();
            }

        }

        // return empty list
        return null;

    }


    @Override
    public Podcast getPodcast(@NotNull String var1) {
        return null;
    }


    @Override
    public List getAllPodcast() {
        return null;
    }


    @Override
    public Channel downloadFeed(@NotNull String var1) {
        return null;
    }


    @Override
    public List getAllEpisodesOfPodcast(@NotNull String var1) {
        return null;
    }


    @Override
    public Episode getEpisode(@NotNull String var1, @NotNull String var2) {
        return null;
    }

    @Override
    public boolean updatePodcast(@NotNull ContentValues var1, @NotNull String var2) {
        return false;
    }

    @Override
    public boolean updateEpisode(@NotNull ContentValues var1, @NotNull Episode var2) {
        return false;
    }

    @Override
    public boolean insertPodcast(@NotNull Podcast var1) {
        return false;
    }

    @Override
    public boolean insertEpisode(@NotNull Episode var1) {
        return false;
    }

    @Override
    public boolean insertEpisodes(@NotNull List<Episode> var1) {
        return false;
    }


}

File: app/src/main/java/getyourcasts/jd/com/getyourcasts/repository/remote/RemoteDataRepository.kt
package getyourcasts.jd.com.getyourcasts.repository.remote

import android.util.Log
import getyourcasts.jd.com.getyourcasts.repository.remote.data.FeedItem
import getyourcasts.jd.com.getyourcasts.repository.remote.data.Podcast
import getyourcasts.jd.com.getyourcasts.repository.remote.network.NetworkHelper

/**
 * Created by chuondao on 7/22/17.
 */


class RemoteDataRepository {

    companion object {
        val TAG = "REMOTE_REPO"
        private var instance: RemoteDataRepository? = null

        /**
         * getter for data instance
         */
        fun getDataInstance(): RemoteDataRepository {
            if (instance == null) {
                instance = RemoteDataRepository()
            }
            return instance!!
        }
    }

    /**
     * search for requested podcast
     */
    fun searchPodcast(title: String): List<Podcast> {
        val results = ArrayList<Podcast>()
        val searchUrl = NetworkHelper.getHelperInstance().searchPodcast(title)
        val http_req = searchUrl.request()
        Log.d(TAG, "Request : $http_req")
        val response = searchUrl.execute()
        /*verify if response is good*/
        if (response.isSuccessful) {
            // parse data here
            if (response.body() != null) {
                return response.body()!!.results
            } else {
                Log.e(TAG, "Failed tor retreive message body")
            }
        }
        return results
    }

    /**
     * download feeds for podcast
     */
    fun fetchEpisodesFromFeedUrl(feedUrl: String): List<FeedItem> {

        // get fetch RSS opt
        val fetchRssOpt = NetworkHelper.getHelperInstance().fetchRss(feedUrl)

        if (fetchRssOpt != null) {
            val req = fetchRssOpt.request()
            Log.d(TAG, "SEND REQUEST $req")
            try {
                val feedResponse = fetchRssOpt.execute()
                if (feedResponse.isSuccessful) {
                    if (feedResponse.body() != null) {
                        return feedResponse.body()!!.channel.items
                    } else {
                        Log.e(TAG, "Failed to fetch Media Item")
                    }
                }
            } catch(e: Exception) {
                Log.e(TAG, "FAILED REQUEST $req. ${e.message} ")
                e.printStackTrace()
            }

        }

        // return empty list
        return ArrayList<FeedItem>()

    }
//
//    override fun getPodcast(podcastId: String): List<Podcast> {
//
//    }
//
//    override fun getAllEpisodesOfPodcast(podcastId: String): List<Episode> {
//        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
//    }
//
//    override fun getEpisode(episodeId: String, podcastId: String): List<Episode> {
//        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
//    }
//
//    override fun updatePodcast(cv: ContentValues, podcastID: String): Long {
//        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
//    }
//
//    override fun updateEpisode(cv: ContentValues, episodeId: String): Long {
//        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
//    }

}